drivers/char/optimsoc-noc.c

```diff
@@ -0,0 +1,436 @@
+
+/* 
+ * Address of memory maped registers.
+ */
+#define REG_NUM_CT (OPTIMSOC_NA_BASE_VADDR + 0x28)
+#define BASE       (OPTIMSOC_NA_BASE_VADDR + 0x100000)
+#define REG_NUMEP  BASE
+#define EP_BASE    BASE + 0x2000
+#define EP_OFFSET  0x2000
+#define REG_SEND   0x0
+#define REG_RECV   0x0
+#define REG_ENABLE 0x4
+/**
+ * Retrieves a word at a target address.
+ */
+#define REG32(x) (*((uint32_t *)(x)))
+
+/*
+ * Buffer size (in bytes).
+ */
+#define OPTIMSOC_NA_BUFFER_SIZE 32
+
+/*
+ * Success.
+ */
+#define SUCCESS 0
+
+/*
+ * Failure
+ */
+#define FAILURE -1
+
+/*
+ * Major device number.
+ */
+static int major;
+
+/*
+ * Number of end points.
+ */
+static uint32_t nr_endpoints;
+
+static struct
+{
+	int nopen;                /* Number of processes using this device. */
+	int head;                 /* Buffer head.                           */
+	int tail;                 /* Buffer tail.                           */
+	int data_received;        /* Was any data received?                 */
+	uint32_t *buffer;         /* Input buffer.                          */
+	wait_queue_head_t wqueue; /* Wait queue of processes.               */
+} adapters[NR_ENDPOINTS];
+
+/*
+ * Enables an endpoint.
+ */
+static void enable(unsigned ep)
+{
+	uint32_t *addr;
+
+#if DEBUG
+	printk(KERN_INFO "enable: %x", ep);
+#endif
+
+	addr = (uint32_t *)(EP_BASE + ep*EP_OFFSET + REG_ENABLE);
+
+	*addr = 1;
+}
+
+/*
+ * Sends a word.
+ */
+static void send(unsigned ep, uint32_t word)
+{
+	uint32_t *addr;
+
+#if DEBUG
+	printk(KERN_INFO "send: %x", word);
+#endif
+
+	addr = (uint32_t *)(EP_BASE + ep*EP_OFFSET + REG_SEND);
+
+	*addr = word;
+}
+
+/*
+ * Receives a word.
+ */
+static uint32_t receive(unsigned ep)
+{
+	uint32_t *addr;
+
+	addr = (uint32_t *)(EP_BASE + ep*EP_OFFSET + REG_RECV);
+
+#if DEBUG
+	printk(KERN_INFO "send: %x", *addr);
+#endif
+
+	return (*addr);
+}
+
+
+/*
+ * Handles a NoC IRQ.
+ */
+static irqreturn_t irq_handler(int irq, void *opaque)
+{
+	int i;
+
+	((void) opaque);
+
+	for (i = 0; i < nr_endpoints; i++)
+	{
+		uint32_t word;
+
+		if (adapters[i].nopen <= 0)
+			continue;
+
+		word = receive(i);
+
+		printk(KERN_INFO "receive: %x", word);
+
+		/* Drop packet. */
+		if ((adapters[i].tail + 1)%OPTIMSOC_NA_BUFFER_SIZE == adapters[i].head)
+			break;
+
+		adapters[i].buffer[adapters[i].tail] = word;
+
+		adapters[i].tail = (adapters[i].tail + 1)%OPTIMSOC_NA_BUFFER_SIZE;
+
+		adapters[i].data_received = 1;
+
+		/* Wakeup sleeping processes. */
+		if (adapters[i].data_received)
+			wake_up(&adapters[i].wqueue);
+	}
+
+	return (IRQ_HANDLED);
+}
+
+/*
+ * Opens the NoC device.
+ */
+static int device_open(struct inode *inode, struct file *file)
+{
+	unsigned minor;
+
+	minor = iminor(inode);
+
+	printk(KERN_INFO "%s: open device %d", OPTIMSOC_NA_NAME, minor);
+
+	/* Device already in use. */
+	if (adapters[minor].nopen > 0)
+		return (-EBUSY);
+
+	/* Grab resources. */
+    adapters[minor].buffer = kmalloc(OPTIMSOC_NA_BUFFER_SIZE, GFP_KERNEL);
```

Comment:
nit: indentation. Run through checkpatch.pl in the linux kernel tree to find all those.